### PR TITLE
Add a guide to contributing hooks.

### PR DESCRIPTION
**Pull request by bwoodsend**
As the title suggests, include a guide for contributing hooks in the README. 

This ended up quite a bit longer than I was expecting - by which I mean that I think it's too long. I'd forgotten just how many hoops we have to jump through.

@colin-fsa As you're the most recent person to ask me the *are any of these steps written down somewhere?* question, I'd appreciate your review on this.

### CODE REVIEW

File: README.md
Comment (rokm) [inline README.md:317]: It might be worth mentioning that `git push` works if one adds new commits to the branch, whereas updating existing commits requires `git push --force`.

**Review by rokm — APPROVED** [review]
That's a nice and detailed writeup of all the required steps (and as you said, there *are* a lot of hoops to go through).

File: README.md
Comment (bwoodsend) [reply README.md:317]: Actually, one nasty git operation for noobs which I've forgotten about: Once you've got one hook merged, how to update your fork to match upstream so that you can do a second?
```
git checkout master
git pull git@github.com:pyinstaller/pyinstaller-hooks-contrib.git
```
does the job but you need to understand how remotes work to know that.

File: README.md
Comment (bwoodsend) [inline README.md:48]: Do you think it would be more sensible to show an https url? I think someone who wants to use ssh is more likely to know to switch the url format than the other way around.

File: README.md
Comment (bwoodsend) [reply README.md:317]: See as we always squash merge (unless someone who knows git well enough to mop up their own history submits multiple hooks in one PR) I was rather hoping that there would never be any need to use `commit --amend` or rebase and therefore never any need to force push. That said, I'm sure someone will somehow prove me wrong the hard way before long.

**Comment by bwoodsend** [comment]
Grr, mental note to never reply to a review comment from the `files changed` tab. Now my replies are all over the place.

File: README.md
Comment (rokm) [reply README.md:317]: > > It might be worth mentioning that git push works if one adds new commits to the branch, whereas updating existing commits requires git push --force.
> 
> See as we always squash merge (unless someone who knows git well enough to mop up their own history submits multiple hooks in one PR) I was rather hoping that there would never be any need to use `commit --amend` or rebase and therefore never any need to force push. That said, I'm sure someone will somehow prove me wrong the hard way before long.

OK, I guess amending commits and force-pushing do indeed fall out of scope here.

Comment (rokm) [reply README.md:317]: > Actually, one nasty git operation for noobs which I've forgotten about: Once you've got one hook merged, how to update your fork to match upstream so that you can do a second?
> 
> ```
> git checkout master
> git pull git@github.com:pyinstaller/pyinstaller-hooks-contrib.git
> ```
> 
> does the job but you need to understand how remotes work to know that.

I think for complete newbies, the GitHub interface also offers a way to sync one's fork with upstream. 

But at least mentioning the CLI way (and perhaps giving a reference to more comprehensive tutorial about remotes, pull vs. pull --rebase, etc.) isn't a bad idea, either.

File: README.md
Comment (rokm) [reply README.md:94]: I'll pass judgment on that one, because I've seen enough `os.path` arithmetic around here that I know exactly what it means...

File: README.md
Comment (rokm) [reply README.md:48]: It might be; if nothing else, GitHub also offers https url by default. Plus, if someone had set keys for ssh access, they will probably know to switch from https url to ssh one.

**Comment by bwoodsend** [comment]
@rokm What do you want to do about the `requirements-libraries.txt`? I've generally (albeit inconsistently) been choosing not to add new libraries in because I didn't like our old per pull request CI just getting longer and longer but now that it's scheduled weekly I reckon we should be adding them in. Possibly even let Pyup loose on it?

File: README.md
Comment (bwoodsend) [reply README.md:318]: I've pointed out the **Fetch upstream** button Github now has. I thought adding commands in there would make it look more complicated than it is.

**Comment by rokm** [comment]
> @rokm What do you want to do about the `requirements-libraries.txt`? I've generally (albeit inconsistently) been choosing not to add new libraries in because I didn't like our old per pull request CI just getting longer and longer but now that it's scheduled weekly I reckon we should be adding them in. Possibly even let Pyup loose on it?

I think the only value of having libraries in there is if we plan to use our weekly CI runs as early-warning breakage detector for new versions of those libraries. But for that to work, we'll need Pyup to keep updating the pinned versions. (Although Pyup's PRs could also trigger CI, so we don't blindly repin to incompatible versions?).

**Comment by bwoodsend** [comment]
> I think the only value of having libraries in there is if we plan to use our weekly CI runs as early-warning breakage detector for new versions of those libraries. But for that to work, we'll need Pyup to keep updating the pinned versions. (Although Pyup's PRs could also trigger CI, so we don't blindly repin to incompatible versions?).

That is roughly what I have in mind. I can set our old per pull request CI/CD workflows to only run on pull requests from the pyup bot which should more or less achieve what we're after. My only issue with Pyup is that it doesn't respect our supported Python versions so there can be a bit of hairiness as packages fail to install on recently dropped Pythons but it's a minor issue.

File: README.md
Comment (colin-fsa) [inline README.md:222]: Needs to be updated, there's no billy-the-buffalo on the line above.

File: README.md
Comment (colin-fsa) [inline README.md:229]: You need to click on 'Oneshot test' on the left to have the run button appear. I'd clarify not to click the 'New Workflow' button since that was my best guess at the next step after I couldn't find the run button.

File: README.md
Comment (colin-fsa) [inline README.md:233]: Adding this to look at later - I believe the pyinstaller dev guide specifies to work on your fork's master branch, presumably if you're suggesting other branch names it's because this guide recommends branching somewhere.

File: README.md
Comment (colin-fsa) [inline README.md:234]: Maybe note to refer to your own commit's entries in requirements-test-libraries.txt? I just did that and noticed I had missed adding libraries.

File: README.md
Comment (colin-fsa) [inline README.md:245]: Since this is a step-by-step guide, it would be helpful to say before the step that you don't need to follow it. I was following along and got confused about how I was supposed to make the decisions for steps 3 and 4.

File: README.md
Comment (bwoodsend) [reply README.md:229]: The url (having the `workflows/oneshot-test.yml` suffix) should've taken care of that. Did you have to click the *enable actions on this fork* page and that didn't forward you directly to the workflow?

File: README.md
Comment (bwoodsend) [reply README.md:233]: > I believe the pyinstaller dev guide specifies to work on your fork's master branch

If it does then it needs changing. Certainly I think the current maintainers all encourage creating branches. And yes, there's a recommendation to do so [at the top](https://github.com/pyinstaller/pyinstaller-hooks-contrib/pull/130/commits/5ac583f7e3f48ef142655ffd6c325a8020c81493#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R53-R58).

File: README.md
Comment (bwoodsend) [reply README.md:234]: Sure. First I need to add something about adding to `requirements-test-libraries.txt` to begin with...

File: README.md
Comment (bwoodsend) [reply README.md:245]: What do you mean by you don't need to follow it? Do you just mean that the defaults will work or something else?

File: README.md
Comment (colin-fsa) [inline README.md:238]: Is this right? I did that and got `ERROR: Invalid requirement: 'spacy srsly thinc'`: https://github.com/colin-fsa/pyinstaller-hooks-contrib/runs/2867341158?check_suite_focus=true

File: README.md
Comment (colin-fsa) [reply README.md:233]: Sorry, you're right. Hit send too quickly.

File: README.md
Comment (colin-fsa) [reply README.md:245]: I think I misread this the first time. It's worded fine.

File: README.md
Comment (colin-fsa) [reply README.md:229]: Yes, that's why

File: README.md
Comment (bwoodsend) [reply README.md:238]: Looks like you wrapped used `"spacy srsly thinc"` instead of just `spacy srsly thinc`?

Comment (colin-fsa) [reply README.md:238]: That seemed like the obvious thing to do since the default for that input is `"numpy==1.19", "numpy<1.18"`

File: README.md
Comment (colin-fsa) [inline README.md:318]: Not to rehash the whole discussion you already had with @rokm, but I think you should at least mention rebasing and force pushing. I had never used either of those things before working on this project and the majority of my time has been spent trying to clean my commits up, so I suspect others would also appreciate some advice or at least training resources.

File: README.md
Comment (colin-fsa) [inline README.md:79]: Just wanted to confirm it was a purposeful decision to not talk about runtime hooks here

File: README.md
Comment (colin-fsa) [inline README.md:80]: Would it be worth giving guidance on how to get started creating a hook if you have absolutely no idea? Like:
```
A common technique for creating a hook is to get the hook working as quickly as possible, then narrow down the minimal logic to keep it working:

- [ ] Write a test, ie. test_foo(). If you're not sure how to write a test try _import foo_, or a minimal reproduceable example of the code that caused you to determine the hook is needed
- [ ] Run the test and ensure it fails: _pytest -k test_foo_ 
- [ ] Create a new hook file and populate it with:
     _from PyInstaller.utils.hooks import collect_all_
    _datas, binaries, hiddenimports = collect_all("foo")_
- [ ] Rerun your test
    If it fails: Additional work will be needed. If you require assistance please feel free to submit a draft and ask for help.
- [ ] If the test passes, replace your existing hook with:
    _datas = copy_metadata('foo')
    binaries = collect_dynamic_libs('foo')
    hiddenimports = collect_submodules('foo')_
- [ ] If the tests still passes, determine if any of these lines can be removed to create a minimal hook
```

File: README.md
Comment (bwoodsend) [reply README.md:39]: See normally I'd still rather they come as separate PRs which just cross reference each other. It's mostly just so that we can squash merge each one and therefore is doesn't matter if the PR author makes a mess of git history.

File: README.md
Comment (bwoodsend) [reply README.md:238]: Yeah I see where you're coming from. It does appear to imply that each job should be quoted. The quotes are only needed to stop bash from interpreting `<` as a redirection. Other than that, they should never be needed. I'm thinking perhaps I could have it escape those characters itself so that there are no quotes involved at all...

File: README.md
Comment (bwoodsend) [reply README.md:80]: I'd quite like something like that on the [page about hooks in the docs](https://pyinstaller.readthedocs.io/en/latest/hooks.html) rather than all the drivel that's on that page currently. Then this readme can just reference it.

File: README.md
Comment (Legorooj) [inline README.md:137]: You appear to have duplicated the headers here - this one should be Apache, I believe.

File: README.md
Comment (bwoodsend) [reply README.md]: I like to pretend that runtime hooks don't exist most of the time. 423b95f7b9f8ca626773e5e22424419a4fcb65ea adds a line on runtime hooks.

File: README.md
Comment (bwoodsend) [reply README.md]: Fixed.

File: README.md
Comment (bwoodsend) [reply README.md:231]: Reckon 7f092f00acf50039f6b199a65103a142533658bc would have kept you right the first time?

File: README.md
Comment (wmmc88) [inline README.md:191]: I think the `minversion` in the `importorskip` no longer works since https://github.com/pyinstaller/pyinstaller/pull/6023. Should it mention `@requires` instead?

Comment (bwoodsend) [reply README.md:191]: Yes, you're right. I've been meaning to address that for ages.

**Pull request merged**
